File: daily-coding-challenge/2026/july/js/18_DiceOdds#342.js
/*

Dice Odds

Given a number of six-sided dice to roll and a target sum, return the odds of rolling that sum as a string in the format "1 in X".
  The number of dice will be between 1 and 6.
  The target sum is always achievable with the given number of dice.
  Round "X" to the nearest whole number.

Tests:
  getOdds(1, 5) should return "1 in 6".
  getOdds(2, 4) should return "1 in 12".
  getOdds(3, 10) should return "1 in 8".
  getOdds(4, 7) should return "1 in 65".
  getOdds(5, 26) should return "1 in 111".
  getOdds(6, 35) should return "1 in 7776".

*/

function getOdds(dice, target) {
  const totalOutcomes = Math.pow(6, dice);

  function countWays(diceLeft, currentSum) {
    if (diceLeft === 0) {
      return currentSum === target ? 1 : 0;
    }

    let ways = 0;
    for (let face = 1; face <= 6; face++) {
      ways += countWays(diceLeft - 1, currentSum + face);
    }
    return ways;
  }

  const successfulWays = countWays(dice, 0);

  const x = Math.round(totalOutcomes / successfulWays);

  return `1 in ${x}`;
}

console.log(getOdds(1, 5)); // "1 in 6".
console.log(getOdds(2, 4)); // "1 in 12".
console.log(getOdds(3, 10)); // "1 in 8".
console.log(getOdds(4, 7)); // "1 in 65".
console.log(getOdds(5, 26)); // "1 in 111".
console.log(getOdds(6, 35)); // "1 in 7776".

File: daily-coding-challenge/2026/july/python/18_DiceOdds#342.py
"""

Dice Odds

Given a number of six-sided dice to roll and a target sum, return the odds of rolling that sum as a string in the format "1 in X".
  The number of dice will be between 1 and 6.
  The target sum is always achievable with the given number of dice.
  Round "X" to the nearest whole number.

Tests:
  get_odds(1, 5) should return "1 in 6".
  get_odds(2, 4) should return "1 in 12".
  get_odds(3, 10) should return "1 in 8".
  get_odds(4, 7) should return "1 in 65".
  get_odds(5, 26) should return "1 in 111".
  get_odds(6, 35) should return "1 in 7776".

"""


def get_odds(dice, target):
    total_outcomes = 6**dice

    def count_ways(dice_left, current_sum):
        if dice_left == 0:
            return 1 if current_sum == target else 0

        ways = 0
        for face in range(1, 7):
            ways += count_ways(dice_left - 1, current_sum + face)
        return ways

    successful_ways = count_ways(dice, 0)

    x = round(total_outcomes / successful_ways)

    return f"1 in {x}"


print(get_odds(1, 5))  # "1 in 6".
print(get_odds(2, 4))  # "1 in 12".
print(get_odds(3, 10))  # "1 in 8".
print(get_odds(4, 7))  # "1 in 65".
print(get_odds(5, 26))  # "1 in 111".
print(get_odds(6, 35))  # "1 in 7776".
